README.md

```diff
@@ -1,526 +1,917 @@
+      };
+
+      initialize(initialValue) {
+        console.log(`initializing ${name}`);
+        return initialValue;
+      }
+    };
+  }
 
-  @deprecated method() { }
+  // ...
+}
 
-  otherMethod() { }
+class C {
+  @logged
+  init m() {}
 }
+
+let c = new C();
+// initializing m
+c.m(1);
+// starting m with arguments 1
+// ending m
 ```
 
-To allow the `deprecated` to work on different kinds of class elements, the `kind` field of the context object lets decorators see which kind of syntactic construct they are deprecating. This technique also allows an error to be thrown when the decorator is used in a context where it can't apply--for example, the entire class cannot be marked as deprecated, since there is no way to intercept its access.
+This example roughly "desugars" to the following:
 
-```mjs
-// deprecated.mjs
+```js
+class C {
+  m() {}
+  m = initializeM(this.m);
+}
 
-function wrapDeprecated(fn) {
-  let name = fn.name
-  function method(...args) {
-    console.warn(`call to deprecated code ${name}`);
-    return fn.call(this, ...args);
-  }
-  Object.defineProperty(method, 'name', { value: name, configurable: true })
-  return method;
-}
-
-export function deprecated(element, {kind}) {
-  switch (kind) {
-    case 'method':
-    case 'getter':
-    case 'setter':
-      return wrapDeprecated(element);
-    case 'field': {
-      let { get, set } = element;
-      return { get: wrapDeprecated(get), set: wrapDeprecated(set) };
-    }
-    default: // includes 'class'
-      throw new Error(`Unsupported @deprecated target ${kind}`);
+let {
+  method,
+  initialize: initializeM
+} = logged(
+  C.prototype.m,
+  {
+    kind: "prop",
+    name: "x",
+    isPrivate: false,
+    defineMetadata() { /**/ }
   }
-}
-```
+);
 
-The desugaring here is analogous to the above examples, which show the use of `kind`.
+C.prototype.m = method;
+```
 
-## Method decorators requiring initialization work
+### Metadata
 
-Some method decorators are based on executing code when the class instance is being created. For example:
+Every decorator has the ability to expose metadata related to the decorated value via the `defineMetadata` method on the context object. This method recieves two parameters, a key which must be a valid property key (string/symbol/number), and a value which can be anything.
 
-- A `@on('event')` decorator for methods on classes extending `HTMLElement` which registers that method as an event listener in the constructor.
-- A `@bound` decorator, which does the equivalent of `this.method = this.method.bind(this)` in the constructor. This idiom meets Jordan Harband's goal of being friendlier to monkey-patching than the popular idiom of using an arrow function in a field initializer.
+```js
+const MY_META = Symbol();
 
-We're considering multiple possible options for how to provide for this type of idiom.
+function myMeta(value, context) {
+  context.defineMetadata("my-meta", true);
+  context.defineMetadata(MY_META, 123);
+}
```

Comment:
One small suggestion for this and the following example code which uses it: actually *using* the `value` passed in would substantially clarify how this is intended to work. The rest of the examples I found quite clear; this one I found a bit harder to follow because the value set on the decorated fields wasn't used in any way.